src/osd_field.c

```diff
@@ -0,0 +1,359 @@
+}
+
+/* Field handlers */
+static void
+field_set_type(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: B (backend) */
+	buf_add(buf, get_type(view, /*short_form*/ false));
+}
+
+static void
+field_set_type_short(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: b (backend) */
+	buf_add(buf, get_type(view, /*short_form*/ true));
+}
+
+static void
+field_set_workspace(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: W */
+	buf_add(buf, view->workspace->name);
+}
+
+static void
+field_set_workspace_short(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: w */
+	if (wl_list_length(&rc.workspace_config.workspaces) > 1) {
+		buf_add(buf, view->workspace->name);
+	}
+}
+
+static void
+field_set_win_state(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: s */
+	if (view->maximized) {
+		buf_add(buf, "M");
+	} else if (view->minimized) {
+		buf_add(buf, "m");
+	} else if (view->fullscreen) {
+		buf_add(buf, "F");
+	} else {
+		buf_add(buf, " ");
+	}
+}
+
+static void
+field_set_win_state_all(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: S */
+	buf_add(buf, view->minimized ? "m" : " ");
+	buf_add(buf, view->maximized ? "M" : " ");
+	buf_add(buf, view->fullscreen ? "F" : " ");
+	/* TODO: add always-on-top and omnipresent ? */
+}
+
+static void
+field_set_output(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: O */
+	if (output_is_usable(view->output)) {
+		buf_add(buf, view->output->wlr_output->name);
+	}
+}
+
+static void
+field_set_output_short(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: o */
+	if (wl_list_length(&view->server->outputs) > 1 &&
+			output_is_usable(view->output)) {
+		buf_add(buf, view->output->wlr_output->name);
+	}
+}
+
+static void
+field_set_identifier(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: I */
+	buf_add(buf, get_app_id_or_class(view, /*trim*/ false));
+}
+
+static void
+field_set_identifier_trimmed(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: i */
+	buf_add(buf, get_app_id_or_class(view, /*trim*/ true));
+}
+
+static void
+field_set_title(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: T */
+	buf_add(buf, get_title(view));
+}
+
+static void
+field_set_title_short(struct buf *buf, struct view *view, const char *format)
+{
+	/* custom type identifier: t */
+	buf_add(buf, get_title_if_different(view));
+}
+
+static void
+field_set_custom(struct buf *buf, struct view *view, const char *format)
+{
+	if (!format) {
+		wlr_log(WLR_ERROR, "Missing format for custom window switcher field");
+		return;
+	}
+
+	char fmt[LAB_FIELD_SINGLE_FMT_MAX_LEN];
+	unsigned char fmt_len = 0;
```

Comment:
Possibly a moot point, but `pos` is more obvious to me here compared with `fmt_len`

Reply:
What about `fmt_pos`? I don't really like the `pos` naming usually if there may be multiple "positions", you never really know what position this refers to otherwise.